exp/controllers/awsmachinepool_controller.go

```diff
@@ -561,6 +572,53 @@ func machinePoolToInfrastructureMapFunc(gvk schema.GroupVersionKind) handler.ToR
 	}
 }
 
+// kubeadmConfigToAWSMachinePool maps a KubeadmConfig to its associated
+// AWSMachinePool. If the KubeadmConfig has an OwnerReferences to a MachinePool,
+// and the owning MachinePool has an InfrastructureRef to an AWSMachinePool, the
+// function returns a reconcile request for that AWSMachinePool.
+func (r *AWSMachinePoolReconciler) kubeadmConfigToAWSMachinePool(o handler.MapObject) []ctrl.Request {
```

Comment:
> We watch KubeadmConfig to detect the changes immediately but it adds certain bootstrap provider dependency as discussed on slack.

Agreed. I've reverted the commit with this code, and added a watch on the bootstrap secret; that has no provider dependency. The _test_ I've written does use a kubeadmconfig; I didn't have any quick ideas for creating a "generic" bootstrap configuration object.

Reply:
> How about doing the userData comparison at every reconcile instead?

I would love to remove this additional watch.

The AWSMachinePool will reconcile, and update the Launch Template:
- When the MachinePool is scaled
- Every 10 minutes (or whatever the CAPA manager's sync period is)

If an EC2 Instance is terminated by the user, or by AWS, ASG immediately replaces it. It may use a Launch Template version with an expired token. The machine will not join the cluster, but, importantly, it will pass the ASG health checks, and will not be terminated.

This additional watch ensures that the Launch Template is updated whenever the token is replaced, e.g., because it's expired. Can we deal with this issue a different way, e.g., by configuring a "Readiness" ASG health check that ensures that an instance that does not join the cluster is terminated?

Reply:
As per discussions in [slack thread](https://kubernetes.slack.com/archives/CD6U2V71N/p1619125890184100?thread_ts=1616059118.005600&cid=CD6U2V71N), I think we should check ASG behaviour when an instance belongs to the ASG is deleted. Does it get recreated? If not, then we won't need to watch the secret. Otherwise we should.

Reply:
> we should check ASG behaviour when an instance belongs to the ASG is deleted. Does it get recreated? If not, then we won't need to watch the secret. Otherwise we should.

That is quite complex :sweat_smile: 

How about I move the change in reconcile to a separate PR? This PR will fix MachinePool scaling.

Reply:
I filed #2369. I have saved the relevant commits from this PR in https://github.com/dlipovetsky/cluster-api-provider-aws/tree/dlipovetsky/machinepool-reconcile-bootstrap-config